File: java-library/src/main/java/com/gblninja/encode/EncodeTags.java
package com.gblninja.encode;


import com.gblninja.tag.GblType;
import com.gblninja.tag.Tag;
import com.gblninja.tag.TagHeader;
import com.gblninja.tag.TagWithHeader;
import com.gblninja.tag.type.GblBootloader;
import com.gblninja.tag.type.GblEnd;
import com.gblninja.tag.type.GblEraseProg;
import com.gblninja.tag.type.GblHeader;
import com.gblninja.tag.type.GblMetadata;
import com.gblninja.tag.type.GblProg;
import com.gblninja.tag.type.GblProgLz4;
import com.gblninja.tag.type.GblProgLzma;
import com.gblninja.tag.type.GblSeUpgrade;
import com.gblninja.tag.type.application.GblApplication;
import com.gblninja.tag.type.certificate.GblCertificateEcdsaP256;
import com.gblninja.tag.type.certificate.GblSignatureEcdsaP256;
import com.gblninja.tag.type.encryption.GblEncryptionData;
import com.gblninja.tag.type.encryption.GblEncryptionInitAesCcm;
import com.gblninja.tag.type.version.GblVersionDependency;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.List;
import java.util.zip.CRC32;

public class EncodeTags {
    private static final int TAG_ID_SIZE = 4;
    private static final int TAG_LENGTH_SIZE = 4;

    public static byte[] encodeTags(List<Tag> tags) {
        int totalSize = calculateTotalSize(tags);
        ByteBuffer buffer = ByteBuffer.allocate(totalSize);
        buffer.order(ByteOrder.LITTLE_ENDIAN);

        for (Tag tag : tags) {
            if (!(tag instanceof TagWithHeader)) continue;

            TagWithHeader tagWithHeader = (TagWithHeader) tag;
            buffer.putInt((int) (tagWithHeader.getTagHeader().getId() & 0xFFFFFFFFL));
            buffer.putInt((int) (tagWithHeader.getTagHeader().getLength() & 0xFFFFFFFFL));

            byte[] tagData = generateTagData(tag);
            buffer.put(tagData);
        }

        return buffer.array();
    }

    public static byte[] generateTagData(Tag tag) {
        if (tag instanceof GblHeader) {
            GblHeader header = (GblHeader) tag;
            ByteBuffer buffer = ByteBuffer.allocate(8)
                    .order(ByteOrder.LITTLE_ENDIAN)
                    .putInt((int) (header.getVersion() & 0xFFFFFFFFL))
                    .putInt((int) (header.getGblType() & 0xFFFFFFFFL));
            return buffer.array();
        }

        if (tag instanceof GblBootloader) {
            GblBootloader bootloader = (GblBootloader) tag;
            ByteBuffer buffer = ByteBuffer.allocate((int) bootloader.getTagHeader().getLength())
                    .order(ByteOrder.LITTLE_ENDIAN)
                    .putInt((int) (bootloader.getBootloaderVersion() & 0xFFFFFFFFL))
                    .putInt((int) (bootloader.getAddress() & 0xFFFFFFFFL));
            buffer.put(bootloader.getData());
            return buffer.array();
        }

        if (tag instanceof GblApplication) {
            GblApplication app = (GblApplication) tag;
            ByteBuffer buffer = ByteBuffer.allocate((int) app.getTagHeader().getLength())
                    .order(ByteOrder.LITTLE_ENDIAN)
                    .putInt((int) (app.getApplicationData().getType() & 0xFFFFFFFFL))
                    .putInt((int) (app.getApplicationData().getVersion() & 0xFFFFFFFFL))
                    .putInt((int) (app.getApplicationData().getCapabilities() & 0xFFFFFFFFL))
                    .put((byte) app.getApplicationData().getProductId());

            if (app.getTagHeader().getLength() > 13L) {
                byte[] remainingData = new byte[(int) app.getTagHeader().getLength() - 13];
                System.arraycopy(app.getTagData(), 13, remainingData, 0, remainingData.length);
                buffer.put(remainingData);
            }
            return buffer.array();
        }

        if (tag instanceof GblProg) {
            GblProg prog = (GblProg) tag;
            ByteBuffer buffer = ByteBuffer.allocate((int) prog.getTagHeader().getLength())
                    .order(ByteOrder.LITTLE_ENDIAN)
                    .putInt((int) (prog.getFlashStartAddress() & 0xFFFFFFFFL));
            buffer.put(prog.getData());
            return buffer.array();
        }

        if (tag instanceof GblSeUpgrade) {
            GblSeUpgrade seUpgrade = (GblSeUpgrade) tag;
            ByteBuffer buffer = ByteBuffer.allocate((int) seUpgrade.getTagHeader().getLength())
                    .order(ByteOrder.LITTLE_ENDIAN)
                    .putInt((int) (seUpgrade.getBlobSize() & 0xFFFFFFFFL))
                    .putInt((int) (seUpgrade.getVersion() & 0xFFFFFFFFL));
            buffer.put(seUpgrade.getData());
            return buffer.array();
        }

        if (tag instanceof GblEnd) {
            GblEnd end = (GblEnd) tag;
            System.out.println("Found end tag " + end.getTagHeader().getLength());
            ByteBuffer buffer = ByteBuffer.allocate((int) end.getTagHeader().getLength())
                    .order(ByteOrder.LITTLE_ENDIAN)
                    .putInt((int) (end.getGblCrc() & 0xFFFFFFFFL));
            return buffer.array();
        }

        if (tag instanceof GblMetadata) {
            GblMetadata metadata = (GblMetadata) tag;
            ByteBuffer buffer = ByteBuffer.allocate((int) metadata.getTagHeader().getLength())
                    .order(ByteOrder.LITTLE_ENDIAN)
                    .put(metadata.getMetaData());
            return buffer.array();
        }

        if (tag instanceof GblProgLz4) {
            GblProgLz4 progLz4 = (GblProgLz4) tag;
            ByteBuffer buffer = ByteBuffer.allocate((int) progLz4.getTagHeader().getLength())
                    .order(ByteOrder.LITTLE_ENDIAN)
                    .put(progLz4.getTagData());
            return buffer.array();
        }

        if (tag instanceof GblProgLzma) {
            GblProgLzma progLzma = (GblProgLzma) tag;
            ByteBuffer buffer = ByteBuffer.allocate((int) progLzma.getTagHeader().getLength())
                    .order(ByteOrder.LITTLE_ENDIAN)
                    .put(progLzma.getTagData());
            return buffer.array();
        }

        if (tag instanceof GblEraseProg) {
            GblEraseProg eraseProg = (GblEraseProg) tag;
            ByteBuffer buffer = ByteBuffer.allocate((int) eraseProg.getTagHeader().getLength())
                    .order(ByteOrder.LITTLE_ENDIAN)
                    .put(eraseProg.getTagData());
            return buffer.array();
        }

        if (tag instanceof GblCertificateEcdsaP256) {
            GblCertificateEcdsaP256 cert = (GblCertificateEcdsaP256) tag;
            ByteBuffer buffer = ByteBuffer.allocate((int) cert.getTagHeader().getLength())
                    .order(ByteOrder.LITTLE_ENDIAN)
                    .put((byte) cert.getCertificate().getStructVersion())
                    .put((byte) cert.getCertificate().getFlags())
                    .put((byte) cert.getCertificate().getKey())
                    .putInt((int) (cert.getCertificate().getVersion() & 0xFFFFFFFFL))
                    .put((byte) cert.getCertificate().getSignature());
            return buffer.array();
        }

        if (tag instanceof GblSignatureEcdsaP256) {
            GblSignatureEcdsaP256 sig = (GblSignatureEcdsaP256) tag;
            ByteBuffer buffer = ByteBuffer.allocate((int) sig.getTagHeader().getLength())
                    .order(ByteOrder.LITTLE_ENDIAN)
                    .put((byte) sig.getR())
                    .put((byte) sig.getS());
            return buffer.array();
        }

        if (tag instanceof GblEncryptionData) {
            GblEncryptionData encData = (GblEncryptionData) tag;
            ByteBuffer buffer = ByteBuffer.allocate((int) encData.getTagHeader().getLength())
                    .order(ByteOrder.LITTLE_ENDIAN)
                    .put(encData.getEncryptedGblData());
            return buffer.array();
        }

        if (tag instanceof GblEncryptionInitAesCcm) {
            GblEncryptionInitAesCcm encInit = (GblEncryptionInitAesCcm) tag;
            ByteBuffer buffer = ByteBuffer.allocate((int) encInit.getTagHeader().getLength())
                    .order(ByteOrder.LITTLE_ENDIAN)
                    .putInt((int) (encInit.getMsgLen() & 0xFFFFFFFFL))
                    .put((byte) encInit.getNonce());
            return buffer.array();
        }

        if (tag instanceof GblVersionDependency) {
            GblVersionDependency versionDep = (GblVersionDependency) tag;
            ByteBuffer buffer = ByteBuffer.allocate(8)
                    .order(ByteOrder.LITTLE_ENDIAN)
                    .putInt((int) (versionDep.getImageType().getValue() & 0xFFFFFFFFL))
                    .put((byte) versionDep.getStatement())
                    .putShort((short) versionDep.getReversed())
                    .putInt((int) (versionDep.getVersion() & 0xFFFFFFFFL));
            return buffer.array();
        }

        return new byte[0];
    }

    private static int calculateTotalSize(List<Tag> tags) {
        int totalSize = 0;
        for (Tag tag : tags) {
            if (tag instanceof TagWithHeader) {
                TagWithHeader tagWithHeader = (TagWithHeader) tag;
                totalSize += TAG_ID_SIZE + TAG_LENGTH_SIZE + tagWithHeader.getTagHeader().getLength();
            }
        }
        return totalSize;
    }

    public static GblEnd createEndTagWithCrc(List<Tag> tags) {
        CRC32 crc = new CRC32();

        for (Tag tag : tags) {
            if (!(tag instanceof TagWithHeader)) continue;

            TagWithHeader tagWithHeader = (TagWithHeader) tag;
            
            byte[] tagIdBytes = ByteBuffer.allocate(4)
                    .order(ByteOrder.LITTLE_ENDIAN)
                    .putInt((int) (tagWithHeader.getTagHeader().getId() & 0xFFFFFFFFL))
                    .array();

            byte[] tagLengthBytes = ByteBuffer.allocate(4)
                    .order(ByteOrder.LITTLE_ENDIAN)
                    .putInt((int) (tagWithHeader.getTagHeader().getLength() & 0xFFFFFFFFL))
                    .array();

            byte[] tagData = generateTagData(tag);

            crc.update(tagIdBytes);
            crc.update(tagLengthBytes);
            crc.update(tagData);
        }

        long endTagId = GblType.END.getValue();
        int endTagLength = TAG_LENGTH_SIZE;

        byte[] endTagIdBytes = ByteBuffer.allocate(TAG_LENGTH_SIZE)
                .order(ByteOrder.LITTLE_ENDIAN)
                .putInt((int) (endTagId & 0xFFFFFFFFL))
                .array();

        byte[] endTagLengthBytes = ByteBuffer.allocate(TAG_LENGTH_SIZE)
                .order(ByteOrder.LITTLE_ENDIAN)
                .putInt(endTagLength)
                .array();

        crc.update(endTagIdBytes);
        crc.update(endTagLengthBytes);

        long crcValue = Integer.toUnsignedLong((int) crc.getValue());
        byte[] crcBytes = ByteBuffer.allocate(TAG_LENGTH_SIZE)
                .order(ByteOrder.LITTLE_ENDIAN)
                .putInt((int) (crcValue & 0xFFFFFFFFL))
                .array();

        return new GblEnd(
                new TagHeader(GblType.END.getValue(), TAG_LENGTH_SIZE),
                GblType.END,
                crcValue,
                crcBytes
        );
    }

    public static byte[] encodeTagsWithEndTag(List<Tag> tags) {
        List<Tag> tagsWithoutEnd = tags.stream()
                .filter(tag -> !(tag instanceof GblEnd))
                .collect(java.util.stream.Collectors.toList());

        GblEnd endTag = createEndTagWithCrc(tagsWithoutEnd);
        List<Tag> finalTags = new java.util.ArrayList<>(tagsWithoutEnd);
        finalTags.add(endTag);

        return encodeTags(finalTags);
    }
}

File: gbl/src/main/java/com/onix/gbl/parser/encode/encodeTags.kt
package parser.data.encode

import parser.data.TAG_ID_SIZE
import parser.data.TAG_LENGTH_SIZE
import parser.data.application.ApplicationData
import parser.data.tag.TagInterface
import parser.data.type.*
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.util.zip.CRC32

/**
 * Кодує список тегів у ByteArray відповідно до формату GBL.
 *
 * @param tags Список тегів для кодування
 * @return ByteArray, що містить серіалізовані дані GBL
 */
fun encodeTags(tags: List<TagInterface>): ByteArray {
    val totalSize = calculateTotalSize(tags)
    val buffer = ByteBuffer.allocate(totalSize)
    buffer.order(ByteOrder.LITTLE_ENDIAN)

    // Обробляємо всі теги послідовно
    for (tag in tags) {
        // Записуємо ID тега (4 байти)
        buffer.putInt(tag.tagHeader.id.toInt())

        // Записуємо довжину payload тега (4 байти)
        buffer.putInt(tag.tagHeader.length.toInt())

        // Генеруємо та записуємо актуальні дані тега на основі його типу
        val tagData = generateTagData(tag)
        buffer.put(tagData)
    }

    return buffer.array()
}

/**
 * Генерує актуальні дані тега залежно від його типу.
 *
 * @param tag Тег для генерації даних
 * @return ByteArray з актуальними даними тега
 */
private fun generateTagData(tag: TagInterface): ByteArray {
    return when (tag) {
        is GblHeader -> {
            val buffer = ByteBuffer.allocate(8)
                .order(ByteOrder.LITTLE_ENDIAN)
                .putInt(tag.version.toInt())
                .putInt(tag.gblType.toInt())
            buffer.array()
        }
        is GblBootloader -> {
            val buffer = ByteBuffer.allocate(tag.tagHeader.length.toInt())
                .order(ByteOrder.LITTLE_ENDIAN)
                .putInt(tag.bootloaderVersion.toInt())
                .putInt(tag.address.toInt())

            // Додаємо дані
            buffer.put(tag.data)
            buffer.array()
        }
        is GblApplication -> {
            val appData = tag.applicationData
            val buffer = ByteBuffer.allocate(tag.tagHeader.length.toInt())
                .order(ByteOrder.LITTLE_ENDIAN)
                .putInt(appData.type.toInt())
                .putInt(appData.version.toInt())
                .putInt(appData.capabilities.toInt())
                .put(appData.productId.toByte())

            // Додаємо залишок даних, якщо вони є
            if (tag.tagHeader.length > 13u) {
                val remainingData = tag.tagData.copyOfRange(13, tag.tagData.size)
                buffer.put(remainingData)
            }
            buffer.array()
        }
        is GblProg -> {
            val buffer = ByteBuffer.allocate(tag.tagHeader.length.toInt())
                .order(ByteOrder.LITTLE_ENDIAN)
                .putInt(tag.flashStartAddress.toInt())

            // Додаємо дані
            buffer.put(tag.data)
            buffer.array()
        }
        is GblSeUpgrade -> {
            val buffer = ByteBuffer.allocate(tag.tagHeader.length.toInt())
                .order(ByteOrder.LITTLE_ENDIAN)
                .putInt(tag.blobSize.toInt())
                .putInt(tag.version.toInt())

            // Додаємо дані
            buffer.put(tag.data)
            buffer.array()
        }
        is GblEnd -> {
            val buffer = ByteBuffer.allocate(4)
                .order(ByteOrder.LITTLE_ENDIAN)
                .putInt(tag.gblCrc.toInt())
            buffer.array()
        }
        is GblMetadata -> {
            tag.metaData
        }
        is GblProgLz4 -> {
            tag.tagData
        }
        is GblProgLzma -> {
            tag.tagData
        }
        is GblEraseProg -> {
            tag.tagData
        }
        else -> {
            tag.tagData
        }
    }
}

/**
 * Обчислює загальний розмір усіх тегів у байтах.
 *
 * @param tags Список тегів
 * @return Загальний розмір у байтах
 */
private fun calculateTotalSize(tags: List<TagInterface>): Int {
    return tags.sumOf { tag ->
        TAG_ID_SIZE + TAG_LENGTH_SIZE + tag.tagHeader.length.toInt()
    }
}

/**
 * Альтернативна версія функції encodeTags з обчисленням CRC.
 * Генерує CRC для кожного тега, якщо це необхідно за форматом GBL.
 *
 * @param tags Список тегів для кодування
 * @param includeCrc Чи включати CRC в кінці кожного тега
 * @return ByteArray, що містить серіалізовані дані GBL з CRC
 */
fun encodeTagsWithCrc(tags: List<TagInterface>, includeCrc: Boolean = false): ByteArray {
    val crcSize = if (includeCrc) 4 else 0

    // Обчислюємо загальний розмір з урахуванням CRC
    val totalSize = tags.sumOf { tag ->
        TAG_ID_SIZE + TAG_LENGTH_SIZE + tag.tagHeader.length.toInt() + crcSize
    }

    val buffer = ByteBuffer.allocate(totalSize)
    buffer.order(ByteOrder.LITTLE_ENDIAN)

    // Обчислюємо загальний CRC для всього файлу
    val fileCrc = CRC32()

    for (tag in tags) {
        val tagIdBytes = ByteBuffer.allocate(4)
            .order(ByteOrder.LITTLE_ENDIAN)
            .putInt(tag.tagHeader.id.toInt())
            .array()

        val tagLengthBytes = ByteBuffer.allocate(4)
            .order(ByteOrder.LITTLE_ENDIAN)
            .putInt(tag.tagHeader.length.toInt())
            .array()

        // Генеруємо актуальні дані тега
        val tagData = generateTagData(tag)

        // Обчислюємо CRC для даного тега
        if (includeCrc) {
            val crc = CRC32()
            crc.update(tagIdBytes)
            crc.update(tagLengthBytes)
            crc.update(tagData)

            // Оновлюємо загальний CRC
            fileCrc.update(tagIdBytes)
            fileCrc.update(tagLengthBytes)
            fileCrc.update(tagData)

            // Записуємо дані тега в буфер
            buffer.put(tagIdBytes)
            buffer.put(tagLengthBytes)
            buffer.put(tagData)

            // Додаємо CRC для тега
            buffer.putInt(crc.value.toInt())
        } else {
            // Якщо CRC не потрібен, просто додаємо дані до буфера
            buffer.put(tagIdBytes)
            buffer.put(tagLengthBytes)
            buffer.put(tagData)

            // Оновлюємо загальний CRC
            fileCrc.update(tagIdBytes)
            fileCrc.update(tagLengthBytes)
            fileCrc.update(tagData)
        }
    }

    // Якщо серед тегів є GblEnd, то замінюємо його CRC на обчислений для всього файлу
    // Це актуально якщо потрібно оновити значення CRC в тезі END
    if (!includeCrc) {
        val endTagIndex = tags.indexOfFirst { it is GblEnd }
        if (endTagIndex != -1) {
            val endTag = tags[endTagIndex] as GblEnd
            val endTagPosition = tags.subList(0, endTagIndex).sumOf {
                TAG_ID_SIZE + TAG_LENGTH_SIZE + it.tagHeader.length.toInt()
            }

            // Позиція, куди записати CRC в буфері
            val crcPosition = endTagPosition + TAG_ID_SIZE + TAG_LENGTH_SIZE

            // Записуємо обчислений CRC в буфер на позицію тега END
            buffer.position(crcPosition)
            buffer.putInt(fileCrc.value.toInt())
        }
    }

    return buffer.array()
}

/**
 * Створює новий об'єкт GblEnd з коректним CRC для заданого списку тегів.
 *
 * @param tags Список тегів (без END тега)
 * @return Новий об'єкт GblEnd з коректним CRC
 */
fun createEndTagWithCrc(tags: List<TagInterface>): GblEnd {
    val crc = CRC32()

    for (tag in tags) {
        val tagIdBytes = ByteBuffer.allocate(4)
            .order(ByteOrder.LITTLE_ENDIAN)
            .putInt(tag.tagHeader.id.toInt())
            .array()

        val tagLengthBytes = ByteBuffer.allocate(4)
            .order(ByteOrder.LITTLE_ENDIAN)
            .putInt(tag.tagHeader.length.toInt())
            .array()

        // Використовуємо актуальні дані тега
        val tagData = generateTagData(tag)

        crc.update(tagIdBytes)
        crc.update(tagLengthBytes)
        crc.update(tagData)
    }

    // Додаємо ID і довжину END тега до CRC
    val endTagId = GblType.END.value.toInt()
    val endTagLength = 4 // Довжина CRC

    val endTagIdBytes = ByteBuffer.allocate(4)
        .order(ByteOrder.LITTLE_ENDIAN)
        .putInt(endTagId)
        .array()

    val endTagLengthBytes = ByteBuffer.allocate(4)
        .order(ByteOrder.LITTLE_ENDIAN)
        .putInt(endTagLength)
        .array()

    crc.update(endTagIdBytes)
    crc.update(endTagLengthBytes)

    // Створюємо байтовий масив для CRC
    val crcValue = crc.value.toInt()
    val crcBytes = ByteBuffer.allocate(4)
        .order(ByteOrder.LITTLE_ENDIAN)
        .putInt(crcValue)
        .array()

    // Створюємо новий END тег
    return GblEnd(
        tagHeader = parser.data.tag.TagHeader(
            id = GblType.END.value.toUInt(),
            length = 4u
        ),
        tagType = GblType.END,
        gblCrc = crcValue.toUInt(),
        tagData = crcBytes
    )
}

/**
 * Кодує список тегів в ByteArray та додає END тег з правильним CRC.
 *
 * @param tags Список тегів (без END тега)
 * @return ByteArray із закодованими тегами та END тегом
 */
fun encodeTagsWithEndTag(tags: List<TagInterface>): ByteArray {
    // Фільтруємо існуючі END теги, щоб додати свій з правильним CRC
    val tagsWithoutEnd = tags.filter { it !is GblEnd }

    // Створюємо END тег з правильним CRC
    val endTag = createEndTagWithCrc(tagsWithoutEnd)

    // Додаємо END тег до списку
    val finalTags = tagsWithoutEnd + endTag

    // Кодуємо всі теги
    return encodeTags(finalTags)
}